modules/ui/fields/sources.js

```diff
@@ -0,0 +1,134 @@
+import { dispatch as d3_dispatch } from 'd3-dispatch';
+import { select as d3_select } from 'd3-selection';
+
+import { t } from '../../core/localizer';
+import { utilGetSetValue, utilNoAuto, utilRebind } from '../../util';
+
+export function uiFieldSources(field, context) {
+    let dispatch = d3_dispatch('change');
+    let items = d3_select(null);
+    let _tags = {};
+    let _selection = d3_select(null);
+    let _pendingChange;
+
+    const mainKey = 'source';
+    const sourceHeader = mainKey + ':';
+
+    // Pre-selected subkey values to show
+    const possibleSourceSubkeys = [{key:'name', value:'Name'}, {key:'url', value:'URL'}, {key:'date', value:'Date'}];
```

Comment:
Is anything using these `value`s? It seems like they can be unset because we explicitly set the placeholder text using the `placeholder` attribute.